### DISCUSSION TITLE
Compilation to WebAssembly with wasi-sdk

### DISCUSSION THREAD

**Discussion by ansemjo** [omnetpp/omnetpp#1241] [Ideas]
Hey there,

I recently attempted to compile OMNeT++ simulations to WebAssembly (specifically the WASI "preview1" / `wasip1` target) by using the [wasi-sdk](https://github.com/WebAssembly/wasi-sdk) as a toolchain. In a first step, I tested how to compile OMNeT++ to static libraries, that don't depend on any system libraries at runtime; and that worked fine. However, since each simulation project is basically an executable that links *against* the OMNeT++ library, the library itself needs to be compiled to a WebAssembly format, that the wasi-sdk toolchain can then link against. And now that's where I am stuck.

Is there any interest to support WebAssembly as a compilation target in the future?

There is already [zigen/omnetpp-wasm](https://github.com/zigen/omnetpp-wasm) but that focuses on using [Emscripten](https://emscripten.org/) to package the entire Qt GUI on a webpage. It's amazing that this works but I'd be more interested in the [standalone mode](https://github.com/emscripten-core/emscripten/wiki/WebAssembly-Standalone) to produce WASI binaries of only the Cmdenv. Maybe I'll just need to look into this again to see if it can be repurposed.

---

It's been a while since I've tested this so here are some (probably incomplete) notes:

1. To build OMNeT++ with static libraries, set in `configure.user`:
```
SHARED_LIBS=no
WITH_QTENV=no
WITH_OSG=no
WITH_OSGEARTH=no
```
Then run `make clean && make -j$(nproc)` as usual.

2. Build your simulation as a static executable:
```
opp_makemake -f --deep -u Cmdenv -L $__omnetpp_root_dir/lib \
    -K CFLAGS=-static -K LDFLAGS=-static
```

3. Attempt to build the simulation with the wasi-sdk:
```
opp_makemake -f --deep -u Cmdenv -L $__omnetpp_root_dir/lib \
    -K CFLAGS=-static -K LDFLAGS=-static \
  # use emulated signals in wasi-sdk:
    -lwasi-emulated-signal -D_WASI_EMULATED_SIGNAL \
  # do not use POSIX coroutines in omnetpp:
    -DUSE_PORTABLE_COROUTINES
```
But that fails because the previously compiled static library cannot be linked with WebAssembly modules.

4. So rebuild OMNeT++ itself with the wasi-sdk. Setting `CXX` and `CC` beforehand didn't work, so I added the SDK to `PATH` and masked those explicitly for configure:
```
export PATH="$WASI_SDK_PATH/bin:$PATH"
CC= CXX= ./configure --host=wasm32 \
    CFLAGS="-D_WASI_EMULATED_SIGNAL -D_WASI_EMULATED_MMAN -D_WASI_EMULATED_GETPID -D_WASI_EMULATED_PROCESS_CLOCKS" \
    LDFLAGS="-lwasi-emulated-signal -lwasi-emulated-mman -lwasi-emulated-getpid -lwasi-emulated-process-clocks"
```
A small dummy enum is needed in `src/common/filelock.h` ([reference](https://nest.pijul.com/mystical.unicat/monorepo:main/Q7P6NO6DTAU3Y.MOGM4BI)):
```
  #if !defined(F_WRLCK)
  enum {
  F_WRLCK,
  F_RDLCK,
  F_GETLK,
  F_SETLK,
  F_SETLKW,
  F_UNLCK
  };
  #endif
```

But now running `make` fails due to various `longjmp` errors that appear to arise due to the use of exceptions. My current understanding is that you could enable experimental support for exceptions in WebAssembly but neither `wasmtime` nor a browser envionment currently supports that anyway. I've also seen that SQLite is linked a few times and while [SQLite now has a dedicated Wasm build](https://sqlite.org/wasm/doc/trunk/index.md), doing the same for an older version that is bundled with the project is probably quite hard.

**Comment by rhornig** [comment]
> Is there any interest to support WebAssembly as a compilation target in the future?

Definitely. I was toying with the idea for a long time to add some kind of support. There are several places where web assembly could be useful (we have a native shared library used by the IDE, that would be great to get rid of :). The other area is indeed creating architecture independent binaries of teh simulation (for easier deployment on clusters?), or simulation redistribution. Providing a full blown QT GUI is a long stretch though. 

If we can define a minimal viable feature set, I can convert this to an 'issue' and work on various problems on the master branch over time (i.e. work with a goal that it would be part of omnet 7)

The embedded SQLite can be updated to a more recent version (but it can also be disabled if needed for this specific target). Exception support is a must however. An other problematic area might be how to support stackfull co-routines (i.e. support for Activity based modules). Do you know anything in the runtime that could be useful for this?

A minimal feature set could be:
- Allow the compilation of statically linked models into webassembly for Cmdenv models. 
- These models could be deployed on any system where the necessary runtime is present.

Explicit non-requirements:
- anything that GUI related (i.e. IDE or the Qtenv)
- dynamically loaded models

**Reply by ansemjo** [reply]
> Definitely.

That's great! :)

> I was toying with the idea for a long time to add some kind of support. There are several places where web assembly could be useful (we have a native shared library used by the IDE, that would be great to get rid of :).

Yeah, I see WebAssembly being used for platform-independent plugin systems more and more, e.g. [recently in Zed](https://zed.dev/blog/zed-decoded-extensions). Replacing the entire core library is probably a different beast though.

> The other area is indeed creating architecture independent binaries of teh simulation (for easier deployment on clusters?), or simulation redistribution.

I'll be frank: that's my primary objective here, yes. I would like to distribute simulation runs by instantiating the simulation binaries with different parameters on an [ad-hoc cluster of volunteer devices](https://github.com/wasimoff/prototype). WebAssembly would make this easy because I need not make any assumptions about the target architecture and WASIp1 binaries can run in any browser with [a small shim](https://github.com/bjorn3/browser_wasi_shim). I haven't looked at [OMNeT's ParSim support](https://doc.omnetpp.org/omnetpp/parsim-api/group__Parsim.html) in-depth yet but I assume that would also benefit from WebAssembly: if the central instance of the simulation model doesn't need to make assumptions about the partitions' architecture and only uses some form of messaging between them, the partitions could just as well be running on a different architecture / in WebAssembly runtimes.

> The embedded SQLite can be updated to a more recent version (but it can also be disabled if needed for this specific target). Exception support is a must however. An other problematic area might be how to support stackfull co-routines (i.e. support for Activity based modules). Do you know anything in the runtime that could be useful for this?

If possible, I'd just disable SQLite entirely then. Right now, I am experimenting with Emscripten some more. As long as the popular WebAssembly runtimes don't support exceptions in WASI natively, I fear that's the only way to compile a runnable Wasm executable from OMNeT++ then. Emscripten does that by pairing the WebAssembly binary with a small JavaScript runtime, which provides the necessary imports to handle jumps and exceptions. Unfortunately that means that you need to *start* the application via the JavaScript runtime, e.g. in NodeJS. While that might be suitable to a browser environment in general, this looses a lot of the portability perks of WebAssembly because you don't have a single self-contained `.wasm` file anymore.

As for the stackfull coroutines, the [threads proposal](https://github.com/WebAssembly/threads/blob/main/proposals/threads/Overview.md) might help here, though it's still considered experimental in `wasi-sdk`. The ["notable limitations" section has a few notes on both exceptions and threads](https://github.com/WebAssembly/wasi-sdk?tab=readme-ov-file#notable-limitations) in their current form.

**Reply by ansemjo** [reply]
Ah, also another note on the goals:

> Allow the compilation of statically linked models into webassembly for Cmdenv models.

It would be great if that was somehow possible without compiling OMNeT++ twice first. Maybe I'm misunderstanding things but as of right now, you first need to compile the `opp_*` tools for your native architecture once because later steps in the Makefile sometimes use these tools. If you were to compile `omnetpp` to WebAssembly immediately, your system usually isn't able to run those produced binaries so the latter steps fail. That's why both I and zigen's project first compile the entire project as an "SDK" and then disable most of the "executable" targets in OMNeT++ when compiling to WebAssembly.

I *think* that Emscripten can handle object files, so at least you wouldn't need to recompile OMNeT++ to WebAssembly first and you could directly compile the simulation itself?

**Reply by torokati44** [reply]
I believe https://github.com/WebAssembly/stack-switching is more in line with what we want as coroutines for activity-based simple modules - they are not concurrently executing threads after all.

And indeed, exception handling is the most important blocker for this. Referencing https://github.com/WebAssembly/wasi-sdk/pull/198 regarding that.

I suppose, as a temporary hack, we could stub the two functions that would normally do the stack unwinding, but some normal behavior of the simulation kernel also relies on C++ exceptions (like ending the simulation for example), so this wouldn't be an actual solution.

Until then, we could go the emscripten + JS engine route, but that doesn't sound ideal either.

**Reply by rhornig** [reply]
> If possible, I'd just disable SQLite entirely then. 
Once I have a little free time, I will add an option in `configure.user` to fully disable the SQLite support, so we don't have to fight that dependency.

> Right now, I am experimenting with Emscripten some more. As long as the popular WebAssembly runtimes don't support exceptions in WASI natively, I fear that's the only way to compile a runnable Wasm executable from OMNeT++ then. 

I agree. I would love to see native WASI modules for at least Cmdenv, but that's indeed depends on exception support, so Emscripten is the way to go (at least for a while). I would be really interested in some performance evaluation between native / Wasm targeted (non-trivial) models. If they are really comparable, we could put more effort into this.

> It would be great if that was somehow possible without compiling OMNeT++ twice first. Maybe I'm misunderstanding things but as of right now, you first need to compile the opp_* tools for your native architecture once because later steps in the Makefile sometimes use these tools.

Correct. Specifically `opp_nedtool` is compiled from c++ which is used during the `.msg` file to `.cpp` generation, but other than that, the rest of the tools are either written in python, bash or they are typically not used during the build process (like result analysis, evenlog filtering etc.)

So that's the only tool that is needed in native form. (ow maybe it would be possible to have a WASI runtime installed and use that to create an executable).

This might be an other actionable item:

- have some mechanism to build only the simulation libraray + Cmdenv and prevent the build of any command line tools

ps: thanks for the provided references. They were interesting.

**Reply by rhornig** [reply]
This seems to be a worthy experiment so I'm converting this to an issue. We will see where this leads us,
